// easycalc/lib/ast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NodeKind;
(function (NodeKind) {
    NodeKind[NodeKind["Const"] = 0] = "Const";
    NodeKind[NodeKind["Identifier"] = 1] = "Identifier";
    NodeKind[NodeKind["Unary"] = 2] = "Unary";
    NodeKind[NodeKind["Binary"] = 3] = "Binary";
})(NodeKind = exports.NodeKind || (exports.NodeKind = {}));
exports.constant = (value, position) => ({
    kind: NodeKind.Const,
    position,
    value
});
exports.identifier = (value, position) => ({
    kind: NodeKind.Identifier,
    position,
    value
});
exports.unary = (operator, value, position) => ({
    kind: NodeKind.Unary,
    position,
    operator,
    value
});
exports.binary = (operator, left, right, position) => ({
    kind: NodeKind.Binary,
    position,
    operator,
    left,
    right
});
function serialize(node) {
    switch (node.kind) {
        case NodeKind.Identifier:
            return `id(${node.value})`;
        case NodeKind.Const:
            return `const(${node.value})`;
        case NodeKind.Unary:
            return `[${node.operator}]<${serialize(node.value)}>`;
        case NodeKind.Binary:
            return `[${node.operator}]<${serialize(node.left)}, ${serialize(node.right)}>`;
    }
}
exports.serialize = serialize;
